### PR TITLE
More progress bar improvements (incl. Jupyter lab support)

### LINKED ISSUE #1087: Rethinking the Jupyter progress bar.
The Jupyter widget API changed quite a bit over the last couple of versions, but seems to get more stable now that they moved it to [its own package](https://ipywidgets.readthedocs.io/en/latest/index.html). We should take another look at the current recommended way to implement and [distribute such widgets](https://ipywidgets.readthedocs.io/en/latest/examples/Custom%20Widget%20-%20Hello%20World.html#More-advanced-uses:-Packaging-and-distributing-Jupyter-widgets) and see whether it makes sense to adopt the recommended approach for Nengo.

### PR DESCRIPTION
**Pull request by jgosmann**
**Motivation and context:**
Jupyter lab disables the execution of Javascript which breaks the current progress bar implementation. Version 0.32 (not released yet, but release candidates are available from PyPI) fixing an updating issue in the VDOM support that allows the implementation of the progress bar without any Javascript. `VdomProgressBar` takes advantage of that and works in Jupyter Lab, it might also work in future versions of Jupyter Notebook and would then allow the removal of `HtmlProgressBar`.

As convenience I added a `VdomOrHtmlProgressBar` that provides both progress bars and the Jupyter client will select the preferred version (usually VDOM if VDOM is supported). This is the new default in clients the support IPython `display`.

There is some sort of code duplication in `VdomProgressBar` and `HtmlProgressBar` as they effectively both construct the same DOM. We could use the VDOM from `VdomProgressBar` and parse it to proper HTML for the `HtmlProgressBar`. However, this might be wasted effort if the `HtmlProgressBar` might get removed in the future. While it is not super complicated todo, the CSS attribute names need to converted from camel case to hyphen notation which might be mildly annoying. But I'm open to other opinions.

Furthermore, this adds a warning that gets displayed in Jupyter Notebook versions before 5 (which are not supported).

Addresses #1426. Closes #1087.

**Interactions with other PRs:**
none

**How has this been tested?**
Manually tried in Jupyter Notebook 5.4.1 + 4.4.1, Jupyter Lab 0.32.0rc1.

**How long should this take to review?**

- Average (neither quick nor lengthy)

**Where should a reviewer start?**
I recommend going commit by commit as they are pretty independent.

**Types of changes:**

- New feature (non-breaking change which adds functionality)

**Checklist:**

- [x] I have read the **CONTRIBUTING.rst** document.
- [x] I have updated the documentation accordingly.
- [x] I have included a changelog entry.
- [n/a] I have added tests to cover my changes.
- [x] I have run the test suite locally and all tests passed.

### CODE REVIEW

**Comment by Seanny123** [comment]
I'll test this on my Windows machine, however note there's a few flake8 errors.

**Comment by jgosmann** [comment]
Fixed the static checks and added changelog entries.

**Comment by Seanny123** [comment]
FYI, this also worked on Jupyter Lab 0.32 on Ubuntu 17.10

**Comment by jgosmann** [comment]
fyi: This progress bar not only works in a notebook opened in Jupyter Lab, but also in a console opened in Jupyter Lab! 😃

**Comment by Seanny123** [comment]
I've been using this branch for more than a month now. It works on Ubuntu 18.04 too! Consequently, I strongly recommend this for merging.

File: nengo/utils/progress.py [hunk 1]
```diff
@@ -399,18 +527,56 @@ def _update_unknown_steps(self, progress):
         '''
 
 
+class VdomOrHtmlProgressBar(ProgressBar):
+    """Progress bar using the VDOM or HTML progress bar.
+
+    This progress bar will transmit both representations as port of a MIME
```
Comment (tbekolay) [inline nengo/utils/progress.py]: Just to make sure, `port` should be `part` right?

**Comment by jgosmann** [comment]
You're right that should be "part".

On August 30, 2018 12:45:36 PM EDT, Trevor Bekolay <notifications@github.com> wrote:
>tbekolay approved this pull request.
>
>LGTM; I'll fix the typo in the merge if you can let me know that it is
>indeed a typo @jgosmann.
>
>> @@ -399,18 +527,56 @@ def _update_unknown_steps(self, progress):
>         '''
> 
> 
>+class VdomOrHtmlProgressBar(ProgressBar):
>+    """Progress bar using the VDOM or HTML progress bar.
>+
>+    This progress bar will transmit both representations as port of a
>MIME
>
>Just to make sure, `port` should be `part` right?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/nengo/nengo/pull/1428#pullrequestreview-151084939

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by tbekolay** [comment]
Great thanks!

**Pull request merged**